rfcs/20201027-modular-tensorflow-graph-c-api.md

```diff
@@ -0,0 +1,527 @@
+2. The graph compiler will not communicate with the plugin using a GraphDef based format, but some TBD format, likely based on serialized [MLIR](https://www.tensorflow.org/mlir).
+
+
+## **User Benefit**
+
+This RFC provides a plugin infrastructure for TensorFlow to optimize graph with new custom optimizers, as long as users set up the system properly installing the graph plugin.
+
+## **Design Proposal**
+
+### Design Overview
+
+This RFC is intended to provide a new mechanism for custom graph optimizers, along with C APIs for users to implement and register their own plugable graph optimizers in Grappler.
+The C APIs follows current implementation C++ API, [TF_Buffer](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/c/c_api.h#L109-L113) and related proto files are the interface between proper and plugin.
+When initializing, TensorFlow loads the plugin and registers a new graph optimizer into Grappler. In the [Optimize](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/grappler/optimizers/graph_optimizer.h#L58) function, plugin authors need to deserialize `TF_Buffer` to `plugin::GraphDef` object to do some graph transformations, and serialize the optimized `plugin::GraphDef` object back to `TF_Buffer` as output.
+
+<p align="center">
+ <img src="20201027-modular-tensorflow-graph-c-api/flow.png" height="400"/>
+</p> 
+
+<!-- Here is what plugin authors need to do:
+1. Implementing `P_Optimize` function.
+2. Copying proto files from proper to plugin, implementing `P_MessageToBuffer`/`P_BufferToMessage` functions.
+3. (Optional)Implementing `plugin::util` functions for graph transformation.
+4. Registering the optimizer -->
+
+### Struct/Function/Object Overview
+- Struct
+  - Struct that should be filled by the plugin: `P_OptimizerConfigFns`, `P_RegistrationParams`, `TF_OptimizerBuilder`
+  - Struct that should be filled by the proper: `TF_GrapplerItem`, `TF_GraphProperties`, `TF_FunctionLibraryDefinition`
+- Function
+  - Function that should be implemented by plugin:
+    - Creation/Deletion/Optimization function of optimizer.
+    - `P_MessageToBuffer`, `P_BufferToMessage`: Serialization/Deserialization of objects generated by protobuf.
+    - `TF_InitGraphPlugin`: Optimizer registration.
+    - (Optional)Util function for graph transformation.
+- Object:
+  - Object defined in plugin: [Object](#proto) generated by protobuf
+
+### Usage Overview
+
+* **Graph Optimization function**
+
+  Graph [Optimize](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/grappler/optimizers/graph_optimizer.h#L58) function is the main part that plugin authors need to implement. The C API looks like below. Both input and output graphs are represented by serialized `TF_Buffer` objects:
+  ```cpp
+  void P_Optimize(void* optimizer, TF_Buffer* graph_buf, TF_Buffer* optimized_graph_buf, TF_Status* s);
```

Comment:
Would it make sense to have a struct for `optimizer` instead of `void*`? I am wondering if there would be some properties we would want to pass in the future.

Reply:
This mechanism is following `Compute` function in kernel C API. Here `void* optimizer` is a struct defined in plugin and created in `P_Create` function, it only contains some members that are used in plugin.

> I am wondering if there would be some properties we would want to pass in the future.
I guess what you referred is `TF_OptimizerBuilder`. Currenty create/delete/optimize is included since these are most useful. We can add more if needed.

Reply:
I have already changed and updated the RFC.